### PR TITLE
AMQNET-831 allow acktype modified_failed

### PR DESCRIPTION
**Pull request by Jefwillems**
This pull request adds the ability to acknowledge messages with MODIFIED_FAILED, instead of MODIFIED_FAILED_UNDELIVERABLE. 
A very basic example below:
```csharp
 private void ReceiveMessage(IMessage message)
    {
        var txt = message.Body<string>();

        _logger.LogInformation("message received: {Content}", txt);


        if (txt.Contains('5') && message is NmsMessage msg)
        {
            msg.NmsAcknowledgeCallback.AcknowledgementType = AckType.MODIFIED_FAILED;
        }

        message.Acknowledge();
    }
```

As far as i can see, this method is the only way to acknowledge a failure when using `AcknowledgementMode.IndividualAcknowledge`

I tested this with an artemis broker, and all tests still run fine.

### CODE REVIEW

File: src/NMS.AMQP/Apache-NMS-AMQP.csproj [hunk 1]
```diff
@@ -32,7 +32,7 @@ with the License.  You may obtain a copy of the License at
 
     <PropertyGroup>
         <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
-        <PackageId>Apache.NMS.AMQP</PackageId>
+        <PackageId>Jef.Apache.NMS.AMQP</PackageId>
```
Comment (michaelandrepearce) [inline src/NMS.AMQP/Apache-NMS-AMQP.csproj:35]: -1

File: .github/GitVersion.yml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+assembly-versioning-scheme: MajorMinorPatch
```
Comment (michaelandrepearce) [inline .github/GitVersion.yml:1]: -1

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
+name: test and publish
```
Comment (michaelandrepearce) [inline .github/workflows/publish.yml:1]: -1 unrelated changes

**Review by michaelandrepearce — COMMENTED** [review]
There's unrelated changes. Also requires a JIRA with commit correctly formed with ref to JIRA.

As this lib is following qpid JMS, can you please direct to same implementation there for reference check.

File: src/NMS.AMQP/Apache-NMS-AMQP.csproj (hunk 1 above)
Comment (Jefwillems) [reply src/NMS.AMQP/Apache-NMS-AMQP.csproj]: didn't realize this would show up here. Done as a temporary fix for a probleem w're having

File: .github/GitVersion.yml (hunk 2 above)
Comment (Jefwillems) [reply .github/GitVersion.yml]: didn't realize this would show up here. Done as a temporary fix for a probleem w're having

File: .github/workflows/publish.yml (hunk 3 above)
Comment (Jefwillems) [reply .github/workflows/publish.yml]: didn't realize this would show up here. Done as a temporary fix for a probleem w're having

**Comment by Jefwillems** [comment]
@michaelandrepearce 
> As this lib is following qpid JMS, can you please direct to same implementation there for reference check.
https://github.com/apache/qpid-jms/blob/main/qpid-jms-client/src/main/java/org/apache/qpid/jms/provider/amqp/AmqpConsumer.java#L323

**Comment by michaelandrepearce** [comment]
@Jefwillems thanks, just need to squash the commits and adjust the commit message to be correctly formed and include the JIRA id.

@Havret any objections here? it looks good to me. On review i noted that we're also missing REJECTED ack type, see here in java variant: https://github.com/apache/qpid-jms/blob/main/qpid-jms-client/src/main/java/org/apache/qpid/jms/provider/amqp/AmqpConsumer.java#L317 Any reason it was left out?

**Comment by Jefwillems** [comment]
> @Jefwillems thanks, just need to squash the commits and adjust the commit message to be correctly formed and include the JIRA id.
> 
> @Havret any objections here? it looks good to me. On review i noted that we're also missing REJECTED ack type, see here in java variant: https://github.com/apache/qpid-jms/blob/main/qpid-jms-client/src/main/java/org/apache/qpid/jms/provider/amqp/AmqpConsumer.java#L317 Any reason it was left out?

Thanks for the feedback, i added the rejected ack type as well. I'll look into squashing and adjusting the message soon

File: src/NMS.AMQP/Provider/Amqp/AmqpConsumer.cs [hunk 4]
```diff
@@ -332,6 +336,10 @@ public void Acknowledge(InboundMessageDispatch envelope, AckType ackType)
                         receiverLink.Modify(message, true, true);
                         RemoveMessage(envelope);
                         break;
+                    case AckType.MODIFIED_FAILED:
+                        receiverLink.Modify(message,true,false);
```
[the remark this answers is not in the document]
Comment (Jefwillems) [reply src/NMS.AMQP/Provider/Amqp/AmqpConsumer.cs]: should be fixed now

**Comment by gemmellr** [comment]
Also for reference, just reaching in and changing the implementation object details directly is not how it is to be used in qpid-jms, the 'ack type' enum is set and converted elsewhere and passed through to that referenced consumer code. We added a JMS_AMQP_ACK_TYPE vendor int property to set on the Message to manipulate the type of disposition actually sent when the Session is acknowledged using that Message objects message.acknowledge() method  (also works when using the individual-acknowledge custom ack mode).

int prop values as:
ACCEPTED = 1;
REJECTED = 2;
RELEASED = 3;
MODIFIED_FAILED = 4;
MODIFIED_FAILED_UNDELIVERABLE = 5;

Testing: https://github.com/apache/qpid-jms/blob/main/qpid-jms-client/src/test/java/org/apache/qpid/jms/integration/AmqpAcknowledgementsIntegrationTest.java#L100-L361

**Comment by Havret** [comment]
Thank you, @gemmellr, for providing the context to our discussion. It's something that Mike and I were previously discussing offline, and I believe it's essential for this change. Without it, the code would be difficult to reach and require casting and digging into implementation details.

**Comment by Jefwillems** [comment]
> @Jefwillems thanks, just need to squash the commits and adjust the commit message to be correctly formed and include the JIRA id.
> 
> @Havret any objections here? it looks good to me. On review i noted that we're also missing REJECTED ack type, see here in java variant: https://github.com/apache/qpid-jms/blob/main/qpid-jms-client/src/main/java/org/apache/qpid/jms/provider/amqp/AmqpConsumer.java#L317 Any reason it was left out?

@Havret i tried squashing but something went wrong, i had to reimplement the tests and the reject outcome. The new commits have the Jira issue in them, but i'm afraid if i try squashing again i'll mess things up. In theory, you should be able to squash commits when merging, if you're merging this.

File: test/Apache-NMS-AMQP-Test/Integration/AmqpAcknowledgmentsIntegrationTest.cs [hunk 5]
```diff
@@ -273,6 +274,219 @@ public void TestAcknowledgeIndividualMessagesAsync()
...
+                for (int i = 0; i < msgCount; i++)
+                {
+                    var message = messages[random.Next(msgCount - i)];
+                    messages.Remove(message);
+
+                    uint deliveryNumber = (uint) message.Properties.GetInt(TestAmqpPeer.MESSAGE_NUMBER) + 1;
+
+                    if (deliveryNumber == 0)
+                    {
+                        testPeer.ExpectDisposition(settled: true, stateMatcher: dispositionMatcherFailed, firstDeliveryId: deliveryNumber, lastDeliveryId: deliveryNumber);
```
Comment (Havret) [inline test/Apache-NMS-AMQP-Test/Integration/AmqpAcknowledgmentsIntegrationTest.cs:329]: This cod in unreachable.

File: test/Apache-NMS-AMQP-Test/Integration/AmqpAcknowledgmentsIntegrationTest.cs [hunk 6]
```diff
@@ -273,6 +274,219 @@ public void TestAcknowledgeIndividualMessagesAsync()
...
```
Comment (Havret) [inline test/Apache-NMS-AMQP-Test/Integration/AmqpAcknowledgmentsIntegrationTest.cs:400]: This is also not reachable.

File: test/Apache-NMS-AMQP-Test/Integration/AmqpAcknowledgmentsIntegrationTest.cs [hunk 7]
```diff
@@ -273,6 +274,219 @@ public void TestAcknowledgeIndividualMessagesAsync()
...
```
Comment (Havret) [inline test/Apache-NMS-AMQP-Test/Integration/AmqpAcknowledgmentsIntegrationTest.cs:471]: This is also unreachable.

File: test/Apache-NMS-AMQP-Test/Integration/AmqpAcknowledgmentsIntegrationTest.cs
Comment (Havret) [inline test/Apache-NMS-AMQP-Test/Integration/AmqpAcknowledgmentsIntegrationTest.cs:1]: I suggest simplifying these tests to focus on a single, straightforward nack without involving randomness. The current approach adds unnecessary complexity and important details, such as the fact that the nack was not being made, are being overlooked.

File: test/Apache-NMS-AMQP-Test/Integration/AmqpAcknowledgmentsIntegrationTest.cs [hunk 8]
```diff
@@ -273,6 +274,219 @@ public void TestAcknowledgeIndividualMessagesAsync()
...
+                        ((NmsMessage) message).NmsAcknowledgeCallback.AcknowledgementType = AckType.MODIFIED_FAILED_UNDELIVERABLE;
```
Comment (Havret) [inline test/Apache-NMS-AMQP-Test/Integration/AmqpAcknowledgmentsIntegrationTest.cs:330]: We should update this to match how it's done in qpid-jms, but we can address it at a later time.

**Review by Havret — CHANGES_REQUESTED** [review]
Thank you for your excellent work @Jefwillems. Your contribution is highly appreciated. However, I have a few remarks that need to be addressed and polished.

File: test/Apache-NMS-AMQP-Test/Integration/AmqpAcknowledgmentsIntegrationTest.cs
Comment (Jefwillems) [reply test/Apache-NMS-AMQP-Test/Integration/AmqpAcknowledgmentsIntegrationTest.cs:1]: You're right, i was stupidly following the example i used to test the change locally. i updated the code to simplify it, and only use nack, instead of both.
